sn_cpu.slurm

```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+#
+#SBATCH --account=watertemp
+#SBATCH --partition=cpu
+#SBATCH --job-name=snakemake
+#SBATCH --time=1-23:59:59
+#SBATCH --ntasks=1
+#SBATCH --cpus-per-task=16
+#SBATCH --output=log/sbatch_all_%a_%A.out
+#SBATCH --error=log/sbatch_all_%a_%A.out
+
+## Initialize work environment
+# conda activate ltls
+# which python
+module load analytics cuda11.3/toolkit/11.3.0
+
+## Run the main job
+
+run_id=test1
+model_id=cpu_a
+log_dir=log/${run_id}_${model_id}
+mkdir -p ${log_dir}
+# From Jeff's blogpost: https://github.com/jsadler2/wdfn-blog/blob/ca63d4dc4bf6aeb6962a58953057a41c9db1ad8a/content/snakemake-ml-experiments.md
+# Also see snakemake profiles: https://snakemake.readthedocs.io/en/stable/executing/cli.html#profiles
+snakemake --cluster "srun -A watertemp -t 1-23:59:59 -p cpu -N 1 -n 1 -c 16 --job-name=${run_id}_${model_id} -e ${log_dir}/slurm-%j.out -o ${log_dir}/slurm-%j.out" --printshellcmds --keep-going --cores all --jobs 1 --rerun-incomplete 3_train/out/mntoha/${run_id}/${model_id}_weights.pt 2>&1 | tee ${log_dir}/run.out
```

Comment:
I've not launched parallel tasks with slurm before, so I'm not sure that I can review this bit - you might want to run it by Jake or Jeff or Jeremy?

If it's running as expected as is, maybe this doesn't even need to be reviewed?

_______________________________

Here are my thoughts (keep that ^ caveat in mind):

When you launch this are you running `sh setup.sh` followed by `sbatch sn_cpu.slurm` (if running on the cpu partition), or `sbatch sn_gpu.slurm` (if running on the gpu portion)?

Here, you've set up a sbatch script in a `.slurm` file. The sbatch script requests an allocation, specifying 8 options (`-A`, `-p`, `-J`, `-t`, `-n`, `-c`, `-o`, and `-e`). Once it has that allocation, it submits a `Snakemake` `--cluster` command that includes a `srun` command that includes a _second_ allocation request, this time specifying 9 options (adding in the `-N` option to specify the number of nodes).

I'm a bit unsure about what's happening here with the double allocation request. It seems like the use of `srun` is right, per [the docs](https://hpcportal.cr.usgs.gov/training/courses/Intro_to_HPC/10_slurm.html#priority-reservations):
> **Execution with srun**
`srun [OPTIONS(0)...] [ : [OPTIONS(N)...]] executable(0) [args(0)...]`
> * Runs a parallel job on cluster managed by Slurm. If necessary, srun will first create a resource allocation in which to run the parallel job.
> * Within an allocation, srun is used to determine what runs on any given subset of resources available
> * Multiple srun commands can be used sequentially or concurrently within a job
> * Failing to use srun within a job allocation will result in the executables being run only on a single node, with an undetermined set of resources. This leads to resource contention and generally bad behavior

...but I'm confused about what that allocation request within the `srun` command is doing. My initial thinking was, given that the `srun` allocation request is identical to the `sbatch` allocation request, is it needed here? But would guess there's a reason you included a second allocation request, and suspect I'm just missing some nuance of slurm. When you run this, what job names, etc. do you see in `squeue`? Are they named `snakemake`, or `test1_cpu_a`? Or is there a single `snakemake` job and then many `test1_cpu_a` jobs?  Also I see that you don't specify the number of nodes in your `sbatch` request, but you do in  your `srun` request. Since the default is 1 task per node and you're only requesting `-n 1` does it not matter?

_To help me think this through, I checked out these other sbatch uses:_
  * In Jeff's blog post, to run tasks in parallel, he submitted a [sbatch command](https://github.com/jsadler2/wdfn-blog/blob/snakemake-dl-experiments/content/snakemake-ml-experiments.md#executing-snakemake-on-many-cores) _within_ a `Snakemake` command

  * For `lake-temperature-neural-networks`, to run tasks in serial, we ran an R script in a [`slurm.sbatch` file](https://code.usgs.gov/hcorson-dosch/lake-temperature-neural-networks/-/blob/master/2_model/src/slurm.sbatch) by [calling `sbatch 2_model/src/slurm.sbatch` in the terminal](https://code.usgs.gov/hcorson-dosch/lake-temperature-neural-networks/-/blob/master/2_model/src/model_tasks.R#L139-142).

Reply:
These are good questions, and while the thing runs now, I don't think that what I've got is perfectly correct. I am using the slurm script in the way you described. Here's how I'm using it:

1. run `./setup.sh` (or, rather, the individual commands in it, since `conda activate` doesn't work inside the script)
2. `sbatch sn_gpu.slurm` or `sbatch sn_cpu.slurm`. 

Here's what I'm hoping it does:
1. Acquire an allocation on gpu/cpu as needed
2. Run the snakemake command on that allocation.
3. Train the model(s) _on that same allocation_ (not sure if this is actually what's happening, but I think it is. Note that I've only tried training one model at a time so far).
4. Use only one node per model, regardless of how many I have allocated to me (no benefit in training one model using multiple nodes).

So, I'm hoping that each `srun` command trains a model on the current allocation, instead of requesting a new allocation. When I run `squeue` I see only one job named snakemake while the model is training. A good sign!

If I use `sbatch` within the snakemake command, like Jeff did, then I think snakemake will make a new allocation request for every model. If the queue time on Tallgrass is long, then the allocation that Snakemake itself runs on might time out before these new allocations start. Maybe one could run snakemake on the head node, but that seems like a bad idea?

If I don't include any --cluster command at all, I'm not sure what would happen. Given the admonition in the `srun` docs that you shared about what happens when you don't use `srun` in an allocation, that seems like a bad idea, too.

I didn't specify the number of nodes in `sbatch` but since `n_tasks=1` and `cpus_per_task=1`, I think I should only get one cpu on one node allocated. I'm thinking of each model that gets trained as one task. I expect to modify `n_tasks` as needed in order to do grid searches for hyperparameter tuning.

I think I should remove `-c 16` from the `srun` command, and let that value be determined by `cpus_per_task` from `sbatch`. I wonder if some of the other flags can be removed, too (`-A`, `-t`, `-p`).

Reply:
> Here's what I'm hoping it does:
> 1. Acquire an allocation on gpu/cpu as needed
> 2. Run the snakemake command on that allocation.
> 3. Train the model(s) on that same allocation (not sure if this is actually what's happening, but I think it is. Note that I've > only tried training one model at a time so far).
> 4. Use only one node per model, regardless of how many I have allocated to me (no benefit in training one model > using multiple nodes).

Thanks for outlining that ^ -- That all sounds good, and it seems like your set up will achieve that, with the possible exception of 3 based on the `srun` options. But I think if you try out training >1 model it _should_ become apparent whether or not the same allocation is being used to submit the `snakemake` command and train the models. I agree that the fact that you only see one job named `snakemake` while a single model is training is a good sign. I _think_ that reflects `srun` recognizing that it doesn't need to create an allocation, since one already exists? But I am curious what job names would be displayed in `squeue` if you trained >1 models. Also, in addition to testing w/ >1 training run I think you could also try dropping some of the options (or all but -J -e and -o) after `srun`, as you suggest.

Reply:
I tried dropping the options `-A`, `-t`, `-p` and `-c`, and everything ran the same.

Reply:
I tried training multiple models as well. The same allocation was used, BUT the models were trained in serial. Uh-oh. It seems that `srun` needs an ampersand after its command in order to run multiple `srun` steps in parallel. Snakemake doesn't make this easy because the input to `--cluster` is basically a prefix, not a suffix. There are two options that I see:

1. Change `srun` to `sbatch`, like Jeff does. That will queue each model up for a new allocation, rather than using the current allocation.
2. In `3_train.smk`, in the rule `train_model`, change `script` to `shell` and hardcode a `srun` command there with an ampersand at the end of it. Then, in `3_train/src/train.py`, change the script part of the file so that it no longer expects snakemake to be an object, and so that it uses command line arguments to determine output file paths, config, etc. instead.

Since option 1 is easier to implement, I'm going that route for now.